### ISSUE TITLE
look into petastorm

### ISSUE THREAD

**Issue by vvksh** [vvksh/learning_stuff#26]
for now, just look at commits for pytorch's petastorm data loader

## context:
most of the data for DL training is in hdfs or some data store; need to read from disk and feed into the model.
that's where petastorm comes in; makes it easy to read parquet file.

created by Uber ATG; seems like optimized for image/bytes data which make for huge row sizes.

to be specific:
- the dataloader needs to return batches of certain batch size for each iteration in a training cycle
- before each epoch, it needs to shuffle the data

## how it reads data?
hypothesis/questions:
1. reads hdfs data using some hdfs library to get parquet
2. decodes parquet using some parquet library
3. in parquet, data is organized by row groups
4. so it must fetch a row group, read bunch of rows, add to batch, if batch full yield, else read another group
5. since its not good to read the whole data (could be huge) in memory to shuffle, what does it do? probably fetch a certain size and shuffle that and return batches from there?? but then how does it handle when it reaches the end of that size? does it fetch some more chunk of data in memory and shuffle that or does it interleave?

### basics:
- Reader (see: https://github.com/vvksh/learning_stuff/issues/26#issuecomment-809986792)
- threadpool
- worker
- RandomShufflingBuffer
- ventilator??

**Comment by vvksh** [comment]
## reader
-> reads from the parquet file
-> `make_reader` returns a Reader instance which has `PyDictReaderWorker` whereas `make_batch_reader` (for general parquet files) returns `Reader` instance with `PyArrowReaderWorker`

**Comment by vvksh** [comment]
###  dataloader code from `pytorch.py`
   -> each `dataloader` instance takes a reader instance -> reader reads a row ~(why? why not read a batch directly from parquet, limitation of underlying library? or network bandwidth?)~  turns out there's a batch reader.
   -> there's a `LoaderBase` abstract class which requires `_iter_impl` method to be implemented.
 
#### pytorch:Dataloader  
   -> looking at `Dataloader` implementation; it first initializes:
    ```
     self._shuffling_buffer = RandomShufflingBuffer(self.shuffling_queue_capacity,
                                                           min_after_retrieve=min_after_dequeue,
                                                           extra_capacity=100000000)
    ```
    -> while its reading row (could be a batch), it adds them to the shuffling queue; and then it checks if it can retrieve a batch, then it gets one row at a time (because of `RandomShufflingBuffer` )  and then adds to `self._batch_acc`; once it reaches `batch_size`, it collates and returns

#### pytorch:BatchedDataloader
-> implemented in [this diff](https://github.com/uber/petastorm/commit/dd87ee630f2f1dfef5200fb854c639134652e209)
-> instead of retrieving row by row from shuffle buffer, it uses pytorch to retrieve a batch directly
-> main logic is in `/pytorch_shuffling_buffer.py` , see the docs, provides nice explanation: https://github.com/uber/petastorm/blob/master/petastorm/reader_impl/pytorch_shuffling_buffer.py#L168
   


     see code https://github.com/uber/petastorm/blob/master/petastorm/pytorch.py#L179
